design/multi-factor-admin-and-step-up.md

```diff
@@ -0,0 +1,582 @@
+possible to require a user to, for example, register an OTP device, it must be possible for the
+administrator to require "register a valid 1st factor" and "register a valid 2nd factor".
+
+To do this, it must be possible to register a required action multiple times, and to configure
+required actions. It must also be possible to remove required actions from the list. The code
+of the required action should define if it can be registered multiple times, and how it can
+be configured. However, all required actions should be able to be removed, and removing a
+required action should also remove it from all users for which it is currently required.
+
+The purpose of this is to be able to have a "register credential" required action in which
+authentication categories can be set. For example, if an admin were to configure a 2nd factor
+with OTP and Fido2, he could then set a "register credential" with the OTP and Fido2
+authentication categories. A user, when confronted with that required action, would only have
+to register one of the two.
+
+
+#### Policies
+
+The password policy for the realm should remain, but the OTP policy should disappear, as a user
+could have several alternative OTP devices with different settings. As such, the information
+about the OTP settings should probably move to information associated with the credential.
+
+### Authentication screens for the user
+
+When the user is redirected towards Keycloak, unless he's using some system that
+bypasses the login actions (e.g. session cookie, kerberos,...) he will arrive at a
+login prompt that corresponds to his authentication category.
+
+The rule for selecting which credential is chosen for a user is
+1. Authentication category that is configured on the client or asked for by the client (see section on step up)
+  1. Credential of the asked for category that the user has selected as preferred (see
+    section on the handling of the users' credentials)
+  1. Credential of the asked for category that is the first in the list of the user
+1. Credential that is among the allowed categories of the authentication factor and that is selected as preferred by the user.
+1. Credential that is first in the list of the user that is of the first authentication category in the authentication flow
+
+The pseudo code for this selection would be:
+```
+if (Client asks for a set of categories OR a set of categories is set for the client) {
+  if (user has a preferred credential in an allowed category) {
+    Ask for a login with that preferred credential
+  } else {
+    Ask for a login with the first credential in the user's list that corresponds to the first allowed category
+  }
+} else if (user has a preferred credential in an allowed category) {
+  Ask for a login with that preferred credential
+} else {
+  Ask for a login with the first credential in the user's list that corresponds to the first allowed category
+}
+```
```

Comment:
Do we need to select one credential from the candidates?
If a user has multiple credentials in an selected authentication category, why don't you try login with all the credentials in the category at once?

Reply:
It may be useful to have such a functionality for some authentication categories, but not others. For example, for FIDO2 it would make sense: you select the category, and any valid FIDO2 token allows you access. 
However, imagine that you have a credential that pushes a request to a smartphone. If the user has multiple such devices recorded, you don't necessarily want to push to all those devices. 
Also, if you have multiple OTP devices recorded (say multiple RSA tokens), here you create a small security issue, as you multiply by the number of recorded OTP devices the chances an attacker has of success by guessing the code.

Reply:
I see. Thank you for clarification.

Reply:
We should not allow trying to match against multiple registered credentials. User should always select a specific credential to validate from the list of configured credentials.

Reply:
How do you think about privacy problem?
In order to let users to select a specific credential, Keycloak need to provide a list of credentials associated with an account to unauthenticated users.
When presenting a list, an attacker may use it as an evidence for a username exists in the directory.

This kind username enumeration risk already exists in a username duplication check of user self sign-up, and many sites do not regard the risk as a problem, and I don't think it is a mistake, but I think we should make an explicit design decision.

Reply:
I agree this is something that is worth highlighting and explicitly mentioning how it is handled. Preventing username enumeration is in direct conflict with usability and smarter authentication flows. I think it's only feasible to prevent username enumeration if there is a required authentication mechanism (password, can't think of anything else realistic here). With a more dynamic flow it is just not possible anymore. Further it's more and more common to do identity first flows (for instance to redirect to a specific IdP for a given domain name, or not ask for password for a user that has configured passwordless).

In summary my take on this is that username enumeration can be prevent if and only if the first page asks for username and password.

Reply:
So the consensus is that we allow username enumeration in order to allow passwordless authentication, and that I'll amend the text to make this explicit?